lib/core/src/recover/handlers/handle_receive_swap.rs

```diff
@@ -0,0 +1,286 @@
+            .clone()
+            .map(|history_tx_id| history_tx_id.txid.to_string());
+        receive_swap.mrh_tx_id = recovered_data
+            .mrh_tx_id
+            .clone()
+            .map(|history_tx_id| history_tx_id.txid.to_string());
+        receive_swap.lockup_tx_id = recovered_data
+            .lockup_tx_id
+            .clone()
+            .map(|history_tx_id| history_tx_id.txid.to_string());
+
+        // Update amounts if we have MRH data
+        if let Some(mrh_amount_sat) = recovered_data.mrh_amount_sat {
+            receive_swap.payer_amount_sat = mrh_amount_sat;
+            receive_swap.receiver_amount_sat = mrh_amount_sat;
+        }
+
+        Ok(())
+    }
+
+    /// Reconstruct Receive Swap tx IDs from the onchain data
+    ///
+    /// The implementation tolerates a `tx_map` that is older than the history in the sense that
+    /// no incorrect data is recovered. Transactions that are missing from `tx_map` are simply not recovered.
+    fn recover_onchain_data(
+        tx_map: &TxMap,
+        swap_id: &str,
+        history: &ReceiveSwapHistory,
+        swap_timestamp: u32,
+    ) -> Result<RecoveredOnchainDataReceive> {
+        // The MRH script history txs filtered by the swap timestamp
+        let mrh_txs: HashMap<Txid, WalletTx> = history
+            .lbtc_mrh_script_history
+            .iter()
+            .filter_map(|h| tx_map.incoming_tx_map.get(&h.txid))
+            .filter(|tx| tx.timestamp.map(|t| t > swap_timestamp).unwrap_or(true))
+            .map(|tx| (tx.txid, tx.clone()))
+            .collect();
+
+        let mrh_tx_id = history
+            .lbtc_mrh_script_history
+            .iter()
+            .find(|&tx| mrh_txs.contains_key::<Txid>(&tx.txid))
+            .cloned();
+
+        let mrh_amount_sat = mrh_tx_id
+            .clone()
+            .and_then(|h| mrh_txs.get(&h.txid))
+            .map(|tx| tx.balance.values().sum::<i64>().unsigned_abs());
+
+        let (lockup_tx_id, claim_tx_id) = determine_incoming_lockup_and_claim_txs(
+            &history.lbtc_claim_script_history,
+            tx_map,
+            swap_id,
+        );
+
+        // Take only the lockup_tx_id and claim_tx_id if either are set,
+        // otherwise take the mrh_tx_id and mrh_amount_sat
+        let recovered_onchain_data = match (lockup_tx_id.as_ref(), claim_tx_id.as_ref()) {
+            (Some(_), None) | (Some(_), Some(_)) => RecoveredOnchainDataReceive {
+                lockup_tx_id,
+                claim_tx_id,
+                mrh_tx_id: None,
+                mrh_amount_sat: None,
+            },
+            _ => RecoveredOnchainDataReceive {
+                lockup_tx_id: None,
+                claim_tx_id: None,
+                mrh_tx_id,
+                mrh_amount_sat,
+            },
+        };
+
+        Ok(recovered_onchain_data)
+    }
+}
+
+/// Helper function for determining lockup and claim transactions in incoming swaps
+fn determine_incoming_lockup_and_claim_txs(
```

Comment:
This function has a duplicate in `handle_chain_receive_swap.rs`. Maybe it could be moved to `handlers/mod.rs` or a new `handlers/utils.rs`?

Reply:
Moved all shared methods.